Permits TokenClassificationOption / DistilBertForTokenClassification to fail gracefully for an invalid configuration.

andyHa opened the pull request:
A production system shouldn't panic, if an invalid config is present. The implementation
is placed in a new method, so that the change is non-breaking.

guillaume-be commented on src/distilbert/distilbert_model.rs:
```diff
@@ -682,30 +682,59 @@ impl DistilBertForTokenClassification {
-            distil_bert_model,
-            classifier,
-            dropout,
+        if let Some(id2label) = &config.id2label {
```
Could raising the error be closer to the problematic field using 
```rust
        let num_labels = config
            .id2label
            .as_ref()
            .ok_or_else(|| RustBertError::InvalidConfigurationError("id2label must be provided in the configuration for sequence classification".to_string()))?
            .len() as i64;
```
This also makes this code block more generic and reusable across the crate

andyHa replied:
Thank's fror the hint - didn't think of that +1

guillaume-be commented on src/distilbert/distilbert_model.rs:
```diff
@@ -682,30 +682,59 @@ impl DistilBertForTokenClassification {
-    where
-        P: Borrow<nn::Path<'p>>,
+    pub fn try_new<'p, P>(p: P, config: &DistilBertConfig) -> Result<DistilBertForTokenClassification, RustBertError>
+        where
```
These edits do not appear when I run `rustfmt`, could you please see if you are using the latest binaries with default settings?

andyHa replied:
Somehow intellij / cargo tricked me here :-/ ...

guillaume-be commented on src/distilbert/distilbert_model.rs:
```diff
@@ -682,30 +682,59 @@ impl DistilBertForTokenClassification {
-    pub fn new<'p, P>(p: P, config: &DistilBertConfig) -> DistilBertForTokenClassification
-    where
-        P: Borrow<nn::Path<'p>>,
+    pub fn try_new<'p, P>(p: P, config: &DistilBertConfig) -> Result<DistilBertForTokenClassification, RustBertError>
```
The API is still undergoing regular re-designs and breaking changes will be pushed for the next release in a few weeks. I would be fine introducing a consistent `new` method for the `SequenceClassificationModels` that would be breaking and avoid adding a `try_new` method to try and keep the API uncluttered. What do you think?

andyHa replied:
Great idea - done.

guillaume-be reviewed (commented):
Thank you for the suggestion @andyHa ,
I have a few minor comments on the DistilBert implementation. This applies to other models, could you please update all instances of `unwrapping/expecting` in these model instantiations for consistency ? If you do not have the capacity please let me know I'd be happy to support.

andyHa commented:
> Thank you for the suggestion @andyHa , I have a few minor comments on the DistilBert implementation. This applies to other models, could you please update all instances of `unwrapping/expecting` in these model instantiations for consistency ? If you do not have the capacity please let me know I'd be happy to support.

@guillaume-be thanks for the feedback - I updated all methods to consistently return a Result and changed all .expect to .ok_or_else(..)?

guillaume-be commented on src/fnet/fnet_model.rs:
```diff
@@ -958,7 +965,7 @@ impl FNetForQuestionAnswering {
     /// # let vs = nn::VarStore::new(device);
     /// # let config = FNetConfig::from_file(config_path);
-    /// let model = FNetForTokenClassification::new(&vs.root(), &config);
+    /// let model = FNetForTokenClassification::new(&vs.root(), &config).unwrap();
```
it looks like this docstring was wrong to start with - could you please fix it with the question answering model instead?

andyHa replied:
Done :)

guillaume-be commented on tests/fnet.rs:
```diff
@@ -227,7 +228,7 @@ fn fnet_for_token_classification() -> anyhow::Result<()> {
     //    Forward pass
     let model_output = no_grad(|| {
-        fnet_model
+        fnet_model.unwrap()
```
can you please be consistent for all tests with the model error handling  (see the changes for `albert` for example):
`let albert_model = AlbertForTokenClassification::new(vs.root(), &config)?;`

andyHa replied:
Done
